app/src/common/shared/org/mozilla/vrbrowser/ui/widgets/TooltipWidget.java

```diff
@@ -0,0 +1,120 @@
+package org.mozilla.vrbrowser.ui.widgets;
+
+import android.content.Context;
+import android.graphics.PointF;
+import android.graphics.Rect;
+import android.view.View;
+import android.view.ViewGroup;
+import android.view.ViewTreeObserver;
+import android.widget.TextView;
+
+import org.mozilla.vrbrowser.R;
+import org.mozilla.vrbrowser.utils.ViewUtils;
+
+public class TooltipWidget extends UIWidget {
+
+    private View mTargetView;
+    private UIWidget mParentWidget;
+    protected TextView mText;
+    private PointF mTranslation;
+    private float mRatio;
+    private int mPaddingH;
+    private int mPaddingV;
+
+    public TooltipWidget(Context aContext) {
+        super(aContext);
+
+        initialize();
+    }
+
+    private void initialize() {
+        inflate(getContext(), R.layout.tooltip, this);
+
+        mText = findViewById(R.id.tooltipText);
+
+        ViewGroup layout = findViewById(R.id.layout);
+        mPaddingH = layout.getPaddingStart() + layout.getPaddingEnd();
+        mPaddingV = layout.getPaddingTop() + layout.getPaddingBottom();
+    }
+
+    @Override
+    protected void initializeWidgetPlacement(WidgetPlacement aPlacement) {
+        aPlacement.visible = false;
```

Comment:
Text looks a bit blurry, specially on the URL bar buttons. I wonder is setting textureScale to 1.0 here can help a bit. Or also trying to add a shadow to the text.

Reply:
I've improved the background for better anti-aliasing, added a shadow and increased the font size and I think the image quality has improved notably.